### ISSUE TITLE
Question: How to manage merge conflicts with MAC signature

### ISSUE THREAD

**Issue by chroto** [mozilla/sops#52]
Hi. Great stuff here.  Putting a PoC together for the team using git as a storage and distribution mechanism.  We are currently using https://www.passwordstore.org/ and I'm attracted to sops for its support for AWS KMS.

What are your recommendations for settling merges since they (almost?) always result in a conflict on the mac signature with neither signature going to be the correct one.

I'm comfortable with `sops --ignore-mac` and `:wq` to re-generate a mac, but this could be a roadblock for adoption since it's additional training to micro-manage the MAC every time there is a merge.

Any recommendations for reducing the scope of this problem?  How do you folks handle this issue?

**Comment by jvehent** [comment]
That's a very valid point, thanks for bringing it up.
Conflict resolution has not been an issue yet, because we manage secrets in small files and it's very unlikely two persons will be modifying the same file at the same time. Generally speaking, I think a lot of small files is better than a single big one.

I have not looked into what `git` provides, but maybe we could invoke a sops command to merge two versions of a file when a conflict occur? That way, sops would verify both MACs, then merge the files and generate a new MAC. It might end up tricky to implement, but worth investigating.

**Comment by twolfson** [comment]
I took a shot at this tonight. I couldn't find anything straightforward with `.gitattributes` but following @jvehent's suggestion, I started on a custom merge configuration for SOPS. Here's a proof of concept demonstrating it handling decryption (which includes MAC validation) of source files and opening the default `mergetool`:

![recording207](https://cloud.githubusercontent.com/assets/902488/14911392/b3820d4e-0dba-11e6-8097-8cf99a0590ca.png)

![recording236](https://cloud.githubusercontent.com/assets/902488/14911395/b69b134a-0dba-11e6-9c81-ff9b5af8e0c1.png)

![recording308](https://cloud.githubusercontent.com/assets/902488/14911397/b80a469c-0dba-11e6-984c-37823507af42.png)

![recording309](https://cloud.githubusercontent.com/assets/902488/14911398/b9604e2e-0dba-11e6-8463-fcb6bbacb869.png)

Code available here:

https://gist.github.com/twolfson/962b1eb776ce9947a09d4924d91fd8b2

There is a snag, we need a SOPS command to replace an entire files contents while maintaining the keys. After that is built, the script should be more/less good to go (maybe port to Python for portability).

Additionally, there is nothing to handle a key rotation. Maybe we should have other commands which dump metadata so we can diff/mergetool those as well?

I should also mention that another short term alternative would be to leverage well tested git hooks for merge/rebase like those in `victorious-git`:

https://github.com/twolfson/victorious-git

**Comment by jvehent** [comment]
Very nice analysis @twolfson ! :+1: 
I only skimmed through your scripts, but my first impression is that we need a bunch of local git configuration, which hurts user experience a bit. Do you think these could be wrapped under a `sops --merge <file>` command which handles all the magical git stuff?

**Comment by twolfson** [comment]
Upon reviewing my code, the `.git/config` + `.gitattributes` are more designed for automatic dependency resolution when running `git merge` (there shouldn't be a mergetool window open, it's all supposed to be automated).

Instead there's `git mergetool` which is used by users when a conflict is encountered. We should be able to use that without any `.git/config` or `.gitattributes`:

```bash
git mergetool --tool=./sops-mergetool.sh example.yaml
```

Unfortunately, I'm not in a good spot to try that out at the moment but will later.

**Comment by twolfson** [comment]
Outside of that I have a couple more notes:

- For replacing an entire file's contents, we should be able to accept input from STDIN but I don't think it works (tried last night)
    - `cat foo.yaml | sops example.yaml`
    - `sops example.yaml < foo.yaml`
- For handling changing meta data like keys, maybe we should pass through meta data to the mergetool as well (maybe via a command like `sops --decrypt-with-meta`). For example:

```yaml
# Inside the 1 column of the mergetool
data:
  foo: bar
sops:
    mac: ABCDEF
    version: 1.7
    pgp:
    -   fp: 1022470DE3F0BC54BC6AB62DE05550BC07FB1A0A
# ...
```

**Comment by jvehent** [comment]
> For replacing an entire file's contents, we should be able to accept input from STDIN but I don't think it works (tried last night) 

Two solutions:
- `sops -e foo.yaml > example.yaml`
- `cp foo.yaml example.yaml && sops -e -i example.yaml`
That omits `--pgp` and `--kms`, but those can be driven through a `.sops.yaml` file.

> For handling changing meta data like keys, maybe we should pass through meta data to the mergetool as well (maybe via a command like sops --decrypt-with-meta).

I think you're looking for `sops --show_master_keys example.yaml` which is already implemented.

**Comment by twolfson** [comment]
With respect to STDIN, is `foo.yaml` raw data (e.g. `foo: bar`)? If so, how does `sops -e` know which keys to use when outputting content?

With respect to `sops --show_master_keys`, is there a way to re-encrypt from that output format? That is convert

```yaml
foo: bar
sops:
    mac: ABCDEF
    version: 1.7
    pgp:
    -   fp: 1022470DE3F0BC54BC6AB62DE05550BC07FB1A0A
```

to 

```yaml
foo: ENC[AES256_GCM,data:9l9h,iv:c/QuxSHYuqV4DWLbMMg1MKEKVbsP14sKFYChHSjxJeU=,tag:EDyqP6ckK0aRhE/XcNt6WQ==,type:str]
sops:
    mac: ENC[AES256_GCM,data:eiKgBSVnGKMIMvLKYees7Q74MXCw+XYSjjOXIWuyJfRuHTNU8dFsKjBhgZBL6RHqu4M5LkzS0MAfg0l9GMV1obnnzD0noht8N/Ge3PQ8QXf2aYYsciDsmBkGksQHC2XWHLF7eheLuicHlyZ2S73i+VAs/gOBywwXR+59x58Vntc=,iv:UlCCFAdOs1HuGrya8gCeoWnW8xj4Q2IarmRxbkwPV5I=,tag:z7l7aYnRqWdfL9wnPXDr8w==,type:str]
    version: 1.7
    pgp:
    -   fp: 1022470DE3F0BC54BC6AB62DE05550BC07FB1A0A
# ...
```

**Comment by twolfson** [comment]
Ah, I have answered my own question in both respects. `-e` accepts data that already has a `sops` key and will use that as its keys

https://github.com/mozilla/sops/blob/f63597f901f50f07ff72452b4bdb485518b85de7/sops/__init__.py#L511-L516

So the answers are:

- No, `foo.yaml` isn't raw data. It is data + a SOPS branch
- Yes, we can re-encrypt `sops --show_master_keys` via `sops -e`

**Comment by twolfson** [comment]
Alright, I explored `git-mergetool`. Unfortunately, we would need to have users set up `.git/config` with a `[mergetool "sops-mergetool"]` to get it working. That being said, the gist is now focused on `git-mergetool` and feels like a good experience (after the `.git/config`):

https://gist.github.com/twolfson/962b1eb776ce9947a09d4924d91fd8b2/1978689a51b7b2f388121fe71a530eae145863fa

![selection_226](https://cloud.githubusercontent.com/assets/902488/14933694/27231764-0e5d-11e6-80bf-d78d3d7d6ce8.png)

![selection_228](https://cloud.githubusercontent.com/assets/902488/14933695/2ae61662-0e5d-11e6-8d48-24f94c79dd0c.png)

![selection_229](https://cloud.githubusercontent.com/assets/902488/14933696/2d5d4262-0e5d-11e6-8c9b-e8528a402e59.png)

![selection_230](https://cloud.githubusercontent.com/assets/902488/14933698/2f9b332c-0e5d-11e6-9672-b917befa18e3.png)

**Comment by twolfson** [comment]
After thinking for a bit, there's no reason why that script has to be used via `git mergetool`. As I learned during iterations, we can generate all the necessary files from the conflicted state:

https://git-scm.com/book/en/v2/Git-Tools-Advanced-Merging#_manual_remerge

```bash
git show :1:example.yaml # Contents of `$BASE`
git show :2:example.yaml # Contents of `$LOCAL`
git show :3:example.yaml # Contents of `$REMOTE`
```

With respect to keeping the tool agnostic (e.g. agnostic to `git`/`hg`/`svg`), I don't think it's plausible for `git` as we need to find its `mergetool` configuration.

**Comment by autrilla** [comment]
> I'm using SOPS via Helm Secrets. Is there a way to turn on `--ignore-mac` via a `sops.yaml` file or some global config?

I think this is a bad idea. You should use `--ignore-mac` reaaaaaally sparingly. It's the only thing that ensures your file hasn't been tampered with by e.g. reordering entries.

> Also, I can't believe this isn't a bigger issue. We ran into this almost immediately, and we're not even that big of an engineering team.

This is probably because of your workflow. I've never had a single merge conflict with our SOPS-encrypted files at Mozilla in over two years, and we're a 10-engineer team. We try to isolate secrets into one file per 'application' and environment.

**Comment by reidab** [comment]
> This is probably because of your workflow. I've never had a single merge conflict with our SOPS-encrypted files at Mozilla in over two years, and we're a 10-engineer team. We try to isolate secrets into one file per 'application' and environment.

I agree that global disabling MAC checking is a bad idea, but I'd be curious to know more about your workflow. Like @cjbottaro, I'm using sops + helm-secrets. 

We currently store secrets files in the same repos as our application code, one file per environment. The place where we most often hit MAC conflicts is when two feature branches each require adding an additional key to the secrets file, and are both merged down to `master`. Git cleanly merges the parallel additions, so the only conflict left to resolve at merge time is the MAC. 

My current workaround, like the OP mentions, is re-saving each conflicted file `sops --ignore-mac`, but this doesn't seem ideal.

**Comment by autrilla** [comment]
@reidab we don't use helm-secrets, so I'm not particularly sure about the viability of this. We currently store secrets file in a separate repo from the application code, and the infrastructure code is in a different repo as well. We do have one file per environment and per application, so that's a similarity between our workflows. We don't hit the issue you hit because our PRs that add secrets are extremely short-lived. We basically handle secrets completely independently from the code that uses them.

I'm not sure how else this could be handled, and I'm definitely open to ideas. I suppose we could write a `sops mergetool` command that does the merge for you, checking the MAC of each of the branches, and then opening the (diffed) decrypted files in your editor and reencrypting the result.

**Comment by cjbottaro** [comment]
@autrilla 
> I think this is a bad idea. You should use --ignore-mac reaaaaaally sparingly. It's the only thing that ensures your file hasn't been tampered with by e.g. reordering entries.

I understand. Is there a way to put that setting in the `.sops.yaml` file though?

**Comment by autrilla** [comment]
> I understand. Is there a way to put that setting in the `.sops.yaml` file though?

Probably not. I don't think we want to make this easy. If you really want it, I'd say you should set up a bash alias or something like that.

**Comment by jvehent** [comment]
I agree with @autrilla . Allowing the user to disable security every time is a bad idea. I like this approach better:

> I suppose we could write a sops mergetool command that does the merge for you, checking the MAC of each of the branches, and then opening the (diffed) decrypted files in your editor and reencrypting the result.

**Comment by ChristianWeissCG** [comment]
How many conflicts you get depends on: 

- your workflow
- frequency of parallel changes
- the number of leave nodes in your yaml file 

Sops was designed which security in mind, not with being compatible with many workflow.

You should always use sops as an interactive editor (`sops filename.yaml`) or modify value by value using `sops --set '["path"]["to"]["key"] "newValue"'`. This ensures that checksum (called "mac") is always correct and also `["sops"]["lastmodified"]` and `["sops"]["version"]` are maintained.

Try to avoid the non-interactive full-file workflow: decryt + change + encrypt, as it introduces in fact a key rotation and puts a lot of noise to a diff, even for not changed values and not changed comments.

This also applies to 3-way-merges. So the 3-way-merge tool should be structure-aware (better then line-based diffs) and it needs to ensure that the "selected" value is applied on top of the encryted version of your "base" file . The 3-way-merge tool should apply a sequence of single-value-patches via `sops --set '["path"]["to"]["key"] "newValue"'` in the background while you are seeing the plain text version of all 3 file versions. The 3-way-merge tool should display only the plain text version without the sops section. That way you should not be asked to solve  conflicts in `["sops"]["mac"]`, `["sops"]["lastmodified"]` and `["sops"]["version"]` - this conflict should be solved under the hood and automatically by the above procedure.

I recomment to not use comments in yaml files. Comments are allowed by the spec but treated by that spec as stepmom children, by declaring them as "presentational detail" (similar to indention and other formating stuff). For that reason many tools like yamldiff or yamlpatch ignore comments completly. While yaml was designed as a format for applications and humans, but sops ecosystem turned out to be focused on applications. Humans are now  stepmom children of the yaml spec. Not using comments would reduce the noise (encrypted comments) and also would help if diff and patch tools ignore comments.

If i really want to have a comment for `'["path"]["to"]["key"] "myValue"'` then i tend to add a new leave node above that key: '["path"]["to"]["key_comment"] "MyCommentGoesHere"'. And as i do not want to have my comment being encrypted i usually use `'["path"]["to"]["key_comment_unencrypted"] "MyCommentGoesHere"'` - this helps when reviewing it at github, gitlab or bitbucket (in the browser) or in my IDE, without forcing me to decrypt all files always. Config options that are not security related also get "_unencrypted" suffix, which also reduces a lot of noise and as they are not part of the mac, changes to them do not trigger a sops-specific conflict.

I hope we will see better tool integration in the future.
Hope to see a `sops --verify` (to verify mac without having a master key; ci server), a `sops --diff` (to create a structure-aware patch file; not line-based like unix diff) and a `sops --applyPatch` to easy review and move config data between feature branches.

BTW: 2-way sopsdiff support via git.textconv is useless on 3-way-merges / conflicts.

**Comment by autrilla** [comment]
> Hope to see a `sops --verify` (to verify mac without having a master key; ci server)

This is not possible, because the MAC is authenticated with the data key, which is not available unless you have a master key.

**Comment by ChristianWeissCG** [comment]
ok, thanks for this detail. 

Having a plain text mac and a signature sounds like a better option as its purpose is to verify that nothing has changed, but not to hide that data. That way a sops command line tool at my build server could verify that a sops-encrypted file is valid, without being a master key owner (envolved in encryption), just need one public key of a person that signed the checksum (should be possible to be signed by more then one person) and need to define trust to at least one of that persons once.

**Comment by autrilla** [comment]
If the MAC isn't authenticated someone can just modify the file by reordering, duplicating, or even swapping values and recomputing the MAC, which defeats its purpose.

We'd be happy to take a patch for some way to improve the merge conflict workflow though.

**Comment by ChristianWeissCG** [comment]
MAC is encrypted with AES_GCM. What i ask for is to not encrypt it, just calculate and then sign that MAC with e.g. GPG (https://www.gnupg.org/gph/en/manual/x135.html):

> In addition to authenticating branches of the tree using keys as additional data, sops computes a MAC on all the values to ensure that no value has been added or removed fraudulently.

This sounds like a simple calculation without encryption involved. A (GPG) signature should be enough. An attacker would need to change the valuse/structure, re-calculate a valid MAC and needs a valid master key from a trusted signee to trick the validation process, by creating a valid signature. But if a master key is no longer private then your encryption is also broken - so it is not worst that current solution. 

Can you provide more details on how the MAC is calculated? (i am not sure if i understand "authenticating branches" correctly)

**Comment by autrilla** [comment]
> MAC is encrypted with AES_GCM. What i ask for is to not encrypt it, just calculate and then sign that MAC with e.g. GPG (https://www.gnupg.org/gph/en/manual/x135.html):

What if the file is only encrypted with AWS KMS? Our master keys can be symmetric keys, which totally defeat your purpose. If you only want support for GPG, you could just sign the file yourself.

What if it's encrypted with both AWS KMS and GPG? Should the MAC only be signed by GPG?

> This sounds like a simple calculation without encryption involved. A (GPG) signature on top of it should be enough. An attacker would need to change the valuse/structure, re-calculate a valid MAC and needs a valid master key from a trusted signee to trick the validation process, by creating a valid signature. But if a master key is no longer private then your encryption is also broken - so it is not worst then current solution.

Sure, it is a simple calculation with no encryption involved. But we still need a way to sign/authenticate the data that works for *all* our master keys. I don't have any formal security training, but my understanding is that this is not possible with symmetric keys.

Also, perhaps this should be discussed in a separate issue? It's not really relevant to how managing merge conflicts should work.

> Can you provide more details on how the MAC is calculated? (i am not sure if i understand "authenticating branches" correctly)

Sure. The relevant code is [here](https://github.com/mozilla/sops/blob/master/sops.go#L295-L348).

**Comment by ChristianWeissCG** [comment]
Thanks for the link. 
True. This should got to a separat ticket: #437

**Comment by ChristianWeissCG** [comment]
"--verify" is not required to implement a "sops auto merge conflict resolver", but it may prevents to commit broken sops files to branches (git hook), which will bring the "sops auto merge conflict resolver" to fail.

**Comment by stpierre** [comment]
> in spite of using very small (per-application, per-environment) files.

So do we. You also have a very small team.

An excellent tool that doesn't work for medium-sized teams or medium-sized files is actually not so excellent. :(

**Comment by ChristianWeissCG** [comment]
As there is no significant development in sops and due to the issue above, i tend to recommend all my teams to use https://www.vaultproject.io/ instead of sops, as sops turned out to be not a good tool for our use case - even with adjusting our workflows to a sops-style.

@stpierre i think the same applies to you and your team. Try to use vault instead.

**Comment by autrilla** [comment]
That there is no significant development is very debatable. We still review and welcome PRs from contributors and have merged a few features lately. Just because a feature on your personal wishlist has not been implemented it doesn't mean that the project is not maintained.

Vault is a completely different beast from sops. If it fits your use case, great! But I don't see how Vault makes this problem any better. In fact, I think it makes it worse. If there's a conflict in Vault, data is just lost, no?

> So do we. You also have a very small team.
> 
> An excellent tool that doesn't work for medium-sized teams or medium-sized files is actually not so excellent. :(

If you have more than 10 people not only accessing, but also editing the same secrets, I think you have a much bigger problem than SOPS not handling merge conflicts well. But again, your use case might be very different than the use cases I imagine for SOPS. Can you go into why this is the case more? It might be good to have a section in the documentation titled something along the lines of "Is SOPS for you?".

**Comment by jvehent** [comment]
As the original author of Sops and manager of the team in charge of it, I can say this: Sops is actively maintained and in heavy use in major services. It will continue being maintained for the foreseeable future.

We implement the features that we need for our own use cases at Mozilla. We also provide feedback and guidance to contributors who wish to implement the features they need for their own use cases., while keeping the project and code base sane and maintainable.

Unlike other secrets management tools, we do not have a business model around Sops. As a result, we do not provide free engineering services. If someone wants a fix for their own use case, they need to implement it themselves. This is usually how open source works, and it doesn't make Sops a lesser tool, but if you were hoping for free beer, sorry, we don't have any.

**Comment by Morriz** [comment]
Is there really no merge conflict problem in the mozilla teams working with sops? I can't believe it. It's inherent to the solution when storing the results in git. Or are you not doing that? Are you working around it in a cumbersome way, breaking git workflows?

**Comment by autrilla** [comment]
We really were doing that, it's not like we gain anything from lying about it. We worked around it by not having a huge file with all our secrets in it, but by splitting secrets into multiple files, separated per service, usually. That, plus pulling before starting to make a change meant I never encountered a merge conflict in my ~2y with the team.

While it'd be great for sops to support this, ISTM that there's also other ways to work around it -- pull before working on something, small files, and quick review turnaround so that merges are quick as well.

**Comment by Perdjesk** [comment]
It is possible to create an alias to ease the regeneration of MAC without having to open interactively an editor.
```bash
alias sops-regen-mac='EDITOR="vim -es +'"'"'norm Go'"'"' +'"'"':wq'"'"'"  sops --ignore-mac'
```
However this will regenerate the MAC even if there is no mismatch in the target file. It might be recommended to only to do it for files that do have mismatches.
```bash
#!/bin/bash

set -o errexit
set -o pipefail
set -o nounset

find . -name '*.enc.yaml' | while read file;
do
    sops -d $file >/dev/null 2>&1 && rc=$? || rc=$?
    # In case of MAC mismatch, then MAC is regenerated
    # See https://github.com/mozilla/sops/blob/v3.6.1/cmd/sops/codes/codes.go#L19
    if [ $rc -eq 51  ] ; then
      echo "Regenerating sops MAC for: $file"
      EDITOR="vim -es +'norm Go' +':wq'"  sops --ignore-mac $file
    fi
done
```


In our team, our git branching workflow handles the MAC mismatch git conflict in pull-requests and it is the task of authors to resolve/regenerate a correct MAC within their PRs. It is for sure a thing more to take care of but it helps to ensure file integrity.

**Comment by mitar** [comment]
In #972 I am proposing that one way to address this might be to have MAC per-value and not per-file.

**Comment by mitar** [comment]
Yes, I noticed this later. See my updated proposal in #972.

**Comment by mitar** [comment]
Given that though, I am not sure what exactly is the threat the per-file MAC is protecting against? That somebody manages to modify the file and put some older (but unknown, encrypted) values for some settings? But that file still has to get deployed without anyone noticing. And also, per-file MAC does not help with revertion of the whole file to some older version. So how insecure really is to always use `sops --ignore-mac`?

**Comment by rubber-ant** [comment]
up

**Comment by jfly** [comment]
Sorry for the bump, but I'm quite interested in the answer to @mitar's question: what is the per-file MAC really protecting against? ["Reordering entries"](https://github.com/getsops/sops/issues/52#issuecomment-444904683) is mentioned above, but that's only something that can happen if someone gains write-access to our repository, which is not a threat we're looking to protect against with sops. We really just want a solution better than committing secrets in plaintext.

In my view of the world: at the point that a malicious actor has the ability to tamper with your repos, they can do all sorts of game-over things (like adding backdoors to your code). If you want to ensure that all the code in a repo was written with good intent, then you should have a robust code submission/code review policy and maybe [consider using signed commits](https://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work).

**Comment by felixfontein** [comment]
MACs authenticate content. Sops encrypted files are not only part of git repositories, but also exist outside of these (for example, some encrypted credentials that are copied from your git repo to a target machine/VM/container). Checking the MAC while decrypting makes sure that these files weren't tampered with *after they were copied out of your repo*. Signed commits won't help for this.

**Comment by jfly** [comment]
*Interesting*. I think that makes sense, but I'm not convinced that's a problem that sops itself should be solving. IIUC, whether this problem exists (and if it's something you care about) really depends on your architeture and how many "steps"/"layers" exist between your git repos that hold sops files and the actual places where the sops binary gets invoked to decrypt those files.

By anology with code: code is also part of git repositories, but also exists outside of them (for example, when building a docker image, .deb file, python package, etc, and uploading that artifact somewhere). Is it absolutely true that we should worry about those artifacts getting tampered with, but that's not something we try to solve by signing the source code: instead, the trusted server/human that reacts to git commits and does the the building/copying is responsible for signing the generated artifact.

**Comment by felixfontein** [comment]
I strongly disagree with that. (I won't comment on this any further in this issue though, since this is getting more off-topic IMO.)

**Comment by mitar** [comment]
> Checking the MAC while decrypting makes sure that these files weren't tampered with after they were copied out of your repo.

Checking MAC is not enough because somebody could still roll-back to an old (possibly compromised) version of the file and MAC checking would still pass. If you worry if files are exactly the version you intend, then simply doing hash comparison of original file and destination file before deploying is good enough and it solves both the attack you are describing as MAC protecting against and roll-back attack.

**Comment by felixfontein** [comment]
Yes, it won't protect about that one. But it does protect against someone modifying the deployed config by copying the TLS private key into the also encrypted "message included in 404 page" field, and sops simply decrypting that without complaining.

**Comment by mitar** [comment]
> Yes, it won't protect about that one. But it does protect against someone modifying the deployed config by copying the TLS private key into the also encrypted "message included in 404 page" field, and sops simply decrypting that without complaining.

No, because individual field is still MACed based on the location in the tree structure.

**Comment by felixfontein** [comment]
Isn't that exactly what I said?

**Comment by mitar** [comment]
Sorry, my understanding is that you are arguing FOR whole-file MACs and were providing arguments what attacks are possible if whole-file MAC is not present. I am arguing FOR field-only MACs (which are already there, because fields are already encrypted with authenticated encryption which takes tree path as part of its additional data to authenticate). And I am arguing that whole-file MACs bring little to none extra utility, while interfering with the user experience of the tool. While security is always a trade off between security and usability, I am claiming that whole-file MACs bring little to none extra security (maybe even just a feeling of security: rollbacks are still possible of the whole file) so that we should at least have an option to have only field-based MACs (at least as opt-in).

**Comment by cgetzen** [comment]
> But it does protect against someone modifying the deployed config by copying the TLS private key into the also encrypted "message included in 404 page" field, and sops simply decrypting that without complaining.

I did an experiment with moving the encrypted data between yaml keys, and was unable to decrypt the file even with `--ignore-mac`. I think this is due to the name of the key being used in the encryption process:

> Additional data is used to guarantee the integrity of the encrypted data and of the tree structure: when encrypting the tree, key names are concatenated into a byte string that is used as AEAD additional data (aad) when encrypting values. We expect that keys do not carry sensitive information, and keeping them in cleartext allows for better diff and overall readability.

[Here](https://github.com/getsops/sops/blob/main/sops.go#L496-L497) is where that is happening.

I think this means that sops files are protected against movement of encrypted values between keys, without needing a MAC.

**Comment by felixfontein** [comment]
Thanks for digging that out, that's indeed correct. I was thinking of the global MAC and not the per-value MAC that's part of the AEAD cipher's output. (That MAC is usually called 'authentication tag'.)

In any case, the global MAC is needed to ensure integrity of the whole file (resp. the encrypted parts, if you use `--mac-only-encrypted`), so you cannot add new keys or replace values (from older versions), or delete keys.

**Comment by felixfontein** [comment]
I was remembering that there was a PR for a sops mergetool for git, but couldn't find it yesterday. I found it while looking at PRs today: #1374. It also seems it was never mentioned in this issue so far.

**Comment by mitar** [comment]
Should this issue be closed now that we have `--mac-only-encrypted`?

**Comment by felixfontein** [comment]
`--mac-only-encrypted` does not help for arbitrary merge conflicts, it only helps for merge conflicts related to changes to unencrypted values/comments. The proper solution for this issue is a mergetool for SOPS files.

**Comment by cgetzen** [comment]
Got it, thanks @felixfontein. I agree that parts could still be removed, added, or reverted. @mitar pointed out that MAC is not enough, as someone could revert the whole file to a previous version. I'll also add that, if a file only has parts of it encrypted, an attacker could create their own sops file that has all encrypted portions removed, but that maintains (or changes) the unencrypted portions.

My organization has found significant value using sops, but this UX issue is a persistent rough edge.

If you feel that the file integrity requirement is still absolutely necessary and should not have the option to be turned off, I have a few straw-man strategies. I'm throwing these out there in part to try to better understand the threat model and qualities you are looking for in a solution, but I would also be happy to work on any of these options in the future if you agree with them.

### **Linked-edge AAD MACs**

This maintains whole-file integrity. To illustrate how this would work, consider the example:
```yaml
a: 1
b: 2
c: 3
d: 4
e: 5
```

The current behavior is for the aad to be the path of the key, which prevents moving encrypted values between keys. Instead of this, if we set the aad to be the path of the current key plus the path of the next key, the encrypted values will be chained together in a way that prevents adding or removing them.

So, to encrypt a's 1, the aad would be set to `a+b`.

This would require adding a `mac` field to the encrypted string in a similar fashion (with overlapping sub-macs of the file) to validate the values.

There are a few caveats to be aware of:

- The merge conflict surface area is now between two keys, which is a reduction compared to a whole file.
- If there are unencrypted keys and `--mac-only-encrypted` is used, then the AAD must point to the next encrypted path, not the next path.
- If there are unencrypted keys and `--mac-only-encrypted` is not used, then the AAD must include the current path, the next encrypted path, and a MAC of all the key/value's between the two encrypted values. The resulting merge conflict surface area is the distance between encrypted keys. 

### **List of 2-pair MACs**
Instead of storing a global MAC, we could MAC each 2-pair (e.g. a+b, b+c, c+d), and store the list of MACs in the file. This would be similar to the above solution, in which the merge conflict surface area is reduced to two keys.

### **--mac-regex value**

Similar to `--encrypted-regex`, this would allow the user to decide which portions of the file need to have integrity. I think the `--mac-only-encrypted` option assumes that integrity is only important for encrypted values, but this may not be the case: e.g. changing an API Hostname setting could cause a security incident.

### **--mac-sections**

This would allow each top-level yaml key to have it's own MAC. An attacker would still be able to revert or delete an entire top-level yaml section.

**Comment by felixfontein** [comment]
I don't think the solution is to add more complexity (complexity is one of the biggest enemies of security), but to simply use a SOPS-aware merge tool (as it has been suggested multiple times in this issue). So far I have seen no arguments why a merge tool will not solve all these problems.

**Comment by cgetzen** [comment]
Great point. I'm not a security engineer, so it's especially interesting to see how experts view these types of proposals.

FWIW, I don't think that a merge tool and trying to reduce merge conflicts are mutually exclusive.

My main concern with a merge tool is that throughput is still limited to 1: Teammates must merge their PRs one-at-a-time and subsequent PRs must pull in the changes. Reverts also become more difficult -- you can no longer use the github or gitlab UI to create reverts if you aren't reverting the most recent change to the file.

Another, more nit-picky concern, is that users must manually configure their git settings to use the merge tool.

Can you help me understand how you weighed the complexity of each proposal? From my perspective, a merge tool is new functionality that widens the tool's surface area, instead of altering (or in a way, reducing, as we're already using `aad` to verify a portion of integrity) the current surface area.

**Comment by cgetzen** [comment]
Here's another case where the MAC does not protect integrity. IMO this belongs in this thread as I believe the chained AAD approach would fix this (though it could also be fixed by adding paths to the MAC):
```yaml
# file v1
old_a: 1
old_b: 2
```
Then, copy the file and edit it with sops:
```yaml
# file v2
new_a: 1
new_b: 2
```
Then, copy the encrypted line from the old file into the new file. It will look something like:
```yaml
# file without integrity
old_a: ENC[AES256_GCM,data:bA==,iv:8heJ3C3Z8QRH/PHrpk8sQ9pL+1aJwI1W8+SlIqB0aSk=,tag:wRLhNoRcWBri1TevS3OVfA==,type:int]
new_b: ENC[AES256_GCM,data:iA==,iv:YD9gh2L64G8eJBIHHyqUf1L/y5PC6373Ek8bPiJ8al0=,tag:5mVOkOkaCCDShYFYubgUsw==,type:int]
```

sops believes the MAC signature is valid in this case.

**Comment by felixfontein** [comment]
That the MAC only depends on the values and not on the keys is a somewhat unfortunate design decision that allows such modifications of the file without tripping the MAC check. I don't know why the MAC is only computed over the values and not the keys.

The best way to avoid this is to compute the MAC over both keys and values, and not just the values. (As you mentioned.)

The Linked-edge AAD MACs you described would also prevent this, but that has other problems though. First, you can simply remove the first key and its value, and the validation will still pass. Then, if you modify a single value in a file, you suddenly get changes in *two* values in the encrypted file. Being able to see in logs with diffs which keys were changed and which keys were not is quite a useful feature of SOPS encrypted config files. This would make it harder to detect which values were actually changed.

The List of 2-pair MACs you described has a similar problem than the Linked-edge AAD MACs when you simply remove the first key + the first 2-pair MAC. Also this has the downside of a large increase in file size if you have a lot of small values, since you get one MAC per value and the MAC is rather large.

> Can you help me understand how you weighed the complexity of each proposal? From my perspective, a merge tool is new functionality that widens the tool's surface area, instead of altering (or in a way, reducing, as we're already using aad to verify a portion of integrity) the current surface area.

A merge tool is new functionality that's totally orthogonal to the encryption/decryption code. No additional complexity is needed in the encryption/decryption code for it. Your proposals increase the complexity in the encryption/decryption code. A complexity increase there is a lot worse IMO than adding more general tools that do not touch the encryption/decryption code.

**Comment by cgetzen** [comment]
Tampering with the first key-value pair can be prevented by using it in the sub-MAC of the last key-value pair, to form a full chain.

Good point about diff visibility becoming more difficult. An individual diff wouldn't be too difficult to infer what changed by ignoring the last value, but using git-blame to discover the last time a value was changed doesn't seem to have a straightforward method (without scripting) to remove false-positives.

Do you think the other option -- a list of MACs which incurs a larger filesize -- could be worth the improved merge throughput? It could be controlled by an optional `--split-macs` flag so an organization could choose which quality is more important for their use-case.

**Comment by cgetzen** [comment]
This continues to be an issue for our growing organization, where we make heavy use of git flow and cherry-picks. A merge tool would not solve the issues we're facing.

Would maintainers accept a pull request that exposes some way to set "--ignore-mac" in [decrypt's](https://pkg.go.dev/github.com/getsops/sops/v3/decrypt) functions? This would remove a large amount of CI issues related to validating configuration.

I believe this would also help solve the terraform provider issue https://github.com/nobbs/terraform-provider-sops/issues/23.
